Add `refineCenters` to replace `MSnbase:::kNeighbours` and `MSnbase:::descendPeak`

sgibb opened the pull request:
This PR introduces a few new functions that I am going to use in
`Spectra::pickPeaks`.

## `valleys`

`valleys` finds the indices of the peak valleys left and right to the peak
(replaces `MSnbase:::.findPeakValley`).


```r
y <- c(1:5, 4:1, 1:5, 4:1)
valleys(y, c(5, 14))
```

```
##      left center right
## [1,]    1      5     9
## [2,]    9     14    18
```

The current implementation (and the original in `MSnbase`) has a known
bug/problem with plateaus, e.g.


```r
y <- c(0, 0, 0, 0, 0, 1:5, 4:1, 0)
valleys(y, 10)
```

```
##      left center right
## [1,]    1     10    15
## left should be 5 here
```

The `localMaxima` function that is used for detection of local minima via `-x`
here, returns the *first* occurence of a local maxima (not the last).
While I could easily change the implementation to return the last local maxima
in a plateau this will result in the same problem but for the right valleys.
Plateaus shouldn't be a real problem in real data beside the baseline (e.g. 0).
So an easy work around would be to call `x[x == 0] <- Inf` before
or more general `x[min(x) == x] <- Inf`.

```r
y <- c(0, 0, 0, 0, 0, 1:5, 4:1, 0)
y[min(y) == y] <- Inf
valleys(y, 10)
```

```
##      left center right
## [1,]    6     10    14
```

But this could maybe create other problems (and `x[min(x) == x] <- Inf` slows
things really down. It needs more than 3/4 of the time of the whole run time of
the `valley` function).

What do you think about that?

## `refineCenters`

`refineCenters` replaces both `MSnbase::kNeighbours` and
`MSnbase::descendPeak`.

In contrast to both `MSnbase` functions `refineCenters` is completely
vectorized. It creates a matrix `c(length(peaks), 2 * k + 1L)`. This improves
the runtime. As always in `R` it is often a compromise between runtime and
memory usage. But even for 1000 peaks and window size of 1000 the resulting
matrix just needs around 8 Mb.

Here some benchmark results for real data:


```r
library("MsCoreUtils")
library("MSnbase")
library("microbenchmark")

data(itraqdata)
sp <- itraqdata[[1]]

mzs <- mz(sp)
ints <- intensity(sp)
pid <- which(localMaxima(ints, hws = 3))

ints2 <- rep(ints, 5)
mzs2 <- seq_along(ints2)
pid2 <- which(localMaxima(ints2, hws = 5))

## define the functions to avoid overhead by `:::`
kNeighbors <- MSnbase:::kNeighbors
descendPeak <- MSnbase:::descendPeak
```

### `kNeighbors`


```r
all.equal(
    kNeighbors(mzs, ints, pid, k = 2)[, 1L],
    refineCenters(mzs, ints, pid, k = 2L, threshold = 0)
)
```

```
## [1] "Mean relative difference: 8.66e-06"
```

```r
options(digits=3)
microbenchmark(
    kNeighbors(mzs, ints, pid, k = 2),
    refineCenters(mzs, ints, pid, k = 2L, threshold = 0),
    kNeighbors(mzs2, ints2, pid2, k = 2),
    refineCenters(mzs2, ints2, pid2, k = 2L, threshold = 0)
)
```

```
## Unit: microseconds
##                                                     expr  min   lq  mean
##                        kNeighbors(mzs, ints, pid, k = 2) 1999 2104  2223
##     refineCenters(mzs, ints, pid, k = 2L, threshold = 0)  199  219   254
##                     kNeighbors(mzs2, ints2, pid2, k = 2) 9056 9638 12684
##  refineCenters(mzs2, ints2, pid2, k = 2L, threshold = 0)  604  678   723
##  median    uq    max neval
##    2197  2310   2933   100
##     244   289    363   100
##    9982 10524 232066   100
##     716   762    974   100
```

```r
microbenchmark(
    kNeighbors(mzs, ints, pid, k = 30),
    refineCenters(mzs, ints, pid, k = 30L, threshold = 0),
    kNeighbors(mzs2, ints2, pid2, k = 30),
    refineCenters(mzs2, ints2, pid2, k = 30L, threshold = 0)
)
```

```
## Unit: microseconds
##                                                      expr   min    lq
##                        kNeighbors(mzs, ints, pid, k = 30)  2172  2294
##     refineCenters(mzs, ints, pid, k = 30L, threshold = 0)   492   548
##                     kNeighbors(mzs2, ints2, pid2, k = 30) 10209 10557
##  refineCenters(mzs2, ints2, pid2, k = 30L, threshold = 0)  2166  2273
##   mean median    uq   max neval
##   2651   2395  2537 14568   100
##    588    585   622   746   100
##  11709  10874 11514 24279   100
##   2930   2366  2521 11855   100
```

```r
microbenchmark(
    kNeighbors(mzs, ints, pid, k = 100),
    refineCenters(mzs, ints, pid, k = 100L, threshold = 0),
    kNeighbors(mzs2, ints2, pid2, k = 100),
    refineCenters(mzs2, ints2, pid2, k = 100L, threshold = 0)
)
```

```
## Unit: milliseconds
##                                                       expr   min    lq
##                        kNeighbors(mzs, ints, pid, k = 100)  2.58  3.01
##     refineCenters(mzs, ints, pid, k = 100L, threshold = 0)  1.33  1.50
##                     kNeighbors(mzs2, ints2, pid2, k = 100) 12.75 14.23
##  refineCenters(mzs2, ints2, pid2, k = 100L, threshold = 0)  6.14  6.99
##   mean median    uq   max neval
##   3.45   3.20  3.40 10.61   100
##   2.00   1.60  1.70  8.05   100
##  16.27  15.04 16.65 25.70   100
##   8.72   7.66  9.74 14.88   100
```

### `descendPeak`


```r
all.equal(
    descendPeak(mzs, ints, pid, k = 30, signalPercentage = 33)[,1L],
    refineCenters(mzs, ints, pid, k = 30L, descending = TRUE, threshold = 0.33)
)
```

```
## [1] TRUE
```

```r
microbenchmark(
    descendPeak(mzs, ints, pid, k = 30, signalPercentage = 33),
    refineCenters(mzs, ints, pid, k = 30L, descending = TRUE, threshold = 0.33),
    descendPeak(mzs2, ints2, pid2, k = 30, signalPercentage = 33),
    refineCenters(mzs2, ints2, pid2, k = 30L, descending = TRUE, threshold = 0.33)
)
```

```
## Unit: microseconds
##                                                                                 expr
##                           descendPeak(mzs, ints, pid, k = 30, signalPercentage = 33)
##          refineCenters(mzs, ints, pid, k = 30L, descending = TRUE, threshold = 0.33)
##                        descendPeak(mzs2, ints2, pid2, k = 30, signalPercentage = 33)
##  refineCenters(mzs2, ints2, pid2, k = 30L, descending = TRUE,      threshold = 0.33)
##    min    lq  mean median    uq    max neval
##   6885  7235  9809   7394  7756 224082   100
##    481   560   596    597   631    717   100
##  31877 34850 37604  35951 41135  48639   100
##   2151  2283  2371   2340  2429   2852   100
```

```r
microbenchmark(
    descendPeak(mzs, ints, pid, k = 30, signalPercentage = 0),
    refineCenters(mzs, ints, pid, k = 30L, descending = TRUE, threshold = 0),
    descendPeak(mzs2, ints2, pid2, k = 30, signalPercentage = 0),
    refineCenters(mzs2, ints2, pid2, k = 30L, descending = TRUE, threshold = 0)
)
```

```
## Unit: microseconds
##                                                                              expr
##                         descendPeak(mzs, ints, pid, k = 30, signalPercentage = 0)
##          refineCenters(mzs, ints, pid, k = 30L, descending = TRUE, threshold = 0)
##                      descendPeak(mzs2, ints2, pid2, k = 30, signalPercentage = 0)
##  refineCenters(mzs2, ints2, pid2, k = 30L, descending = TRUE,      threshold = 0)
##    min    lq  mean median    uq   max neval
##   6765  7296  7740   7441  7771 13625   100
##    512   563   602    598   639   785   100
##  32137 35001 37613  35781 41472 51053   100
##   2032  2304  2397   2355  2498  2782   100
```

jorainer commented:
Thanks @sgibb !
Regarding the known bug in `valleys` - I'm OK to have it, just important to describe that in the function's documentation.

And I like the combination of the two function into a single one - only question would be if we call that `refineCenters` (which reads nicely) or `refineCentroids` (which would be AFAIK the more accurate name).

Regarding speed: impressive! Regarding difference between `kNeighbors` and the new function: what is the reason for that?

jorainer commented on R/refineCenters.R:
```diff
@@ -0,0 +1,143 @@
+#'
+#' @param x `numeric`, i.e. m/z values.
+#' @param y `numeric`, i.e. intensity values.
+#' @param p `numeric`, peak centers.
```
So, `p` is then an m/z value? I assume it is one of the values in `x`? Would it make sense that `p` would be an `integer` defining where in `x` the peaks are?

sgibb replied:
`p` is the peak index. Will change the documentation.

jorainer commented on R/refineCenters.R:
```diff
@@ -0,0 +1,143 @@
+#' @param p `numeric`, peak centers.
+#' @param k `integer(1)`, number of values left and right of the peak that
+#'  should be considered in the weighted mean calculation.
+#' @param threshold `double`, ratio of the maximal peak intensity. Just values
```
Please use `double(1)` here. Also, I think it would be more appropriate to call this the *proportion* or *percentage* of the maximal peak intensity.

sgibb replied:
Thanks, it will be *proportion*.

jorainer commented on R/refineCenters.R:
```diff
@@ -0,0 +1,143 @@
+#'  should be considered in the weighted mean calculation.
+#' @param threshold `double`, ratio of the maximal peak intensity. Just values
+#'  above are used for the weighted mean calclulation.
+#' @param descending `logical`, if `TRUE` just values between the nearest
```
So I guess you are no longer *walking* down the peak as I did, but use signal between the *peak* and the closest *valley*?

sgibb replied:
That's right.

jorainer commented on R/refineCenters.R:
```diff
@@ -0,0 +1,143 @@
+refineCenters <- function(x, y, p, k = 2, threshold = 0.33, descending = TRUE) {
+    if (!is.numeric(x))
+        stop("'x' has to be a numeric vector.")
+    if (!is.numeric(y) || length(x) != length(y))
```
Why not having also `!is.numeric(x)` here and remove the extra `if (!is.numeric(x))` above?

sgibb replied:
Don't know. Adopted to your suggestion.

jorainer commented on R/valleys.R:
```diff
@@ -0,0 +1,49 @@
+#' v <- valleys(ints, peaks)
+#' segments(mzs[v[, "left"]], 0, mzs[v[, "right"]], col = cols, lwd = 2)
+valleys <- function(x, p) {
+    if (!is.numeric(x))
```
Why not putting the `!is.numeric(x)` and `!is.numeric(y)` into the same `if`?

sgibb replied:
I make `p` an `integer` here, as well. So both test a kept (now testing for `is.numeric(x)` and `is.integer(p)`).

sgibb commented:
> Regarding the known bug in `valleys` - I'm OK to have it, just important to describe that in the function's documentation.

Did that now.

>... `refineCentroids` (which would be AFAIK the more accurate name).

Renamed.

> Regarding difference between `kNeighbors` and the new function: what is the reason for that?

My fault. I wanted to see the difference between using `descending = TRUE` (was the default) and `descending = FALSE`. With `descending = FALSE` there is no difference between `kNeighbors` and `refineCentroids`. I now changed the defaults to `k = 2L, descending = FALSE`, or would you like `k = 30L, descending = TRUE` (I never used refinement).

jorainer commented:
Keeping your default is OK.
